File: sdk/php/src/Providers/MaildropCc.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * maildrop.cc 渠道实现（https://maildrop.cc）
 *
 * GraphQL API，无认证，公共邮箱（任何人可查看任意地址收件箱）。
 * generate 无需 API，直接随机用户名 + @maildrop.cc；
 * getEmails POST https://api.maildrop.cc/graphql，先 inbox 查列表，再逐封 message 补 html 正文。
 */
final class MaildropCc
{
    private const CHANNEL = 'maildrop-cc';
    private const DOMAIN = 'maildrop.cc';
    private const GRAPHQL_URL = 'https://api.maildrop.cc/graphql';

    /** @var array<string,string> */
    private const HEADERS = [
        'Accept' => 'application/json',
        'Content-Type' => 'application/json',
        'Origin' => 'https://maildrop.cc',
        'Referer' => 'https://maildrop.cc/',
        'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
            . '(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
    ];

    private const LOCAL_CHARS = 'abcdefghijklmnopqrstuvwxyz0123456789';

    private static function randomLocal(int $n = 10): string
    {
        $out = '';
        for ($i = 0; $i < $n; $i++) {
            $out .= self::LOCAL_CHARS[random_int(0, strlen(self::LOCAL_CHARS) - 1)];
        }
        return $out;
    }

    private static function mailbox(string $email): string
    {
        $email = trim($email);
        $idx = strpos($email, '@');
        return $idx !== false ? substr($email, 0, $idx) : $email;
    }

    /**
     * 发送 GraphQL 请求并返回解析后的 JSON。
     *
     * @return array<mixed>
     */
    private static function doGraphql(string $query): array
    {
        $resp = HttpClient::post(self::GRAPHQL_URL, self::HEADERS, ['query' => $query]);
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('maildrop-cc: GraphQL 请求失败 ' . $resp->getStatusCode());
        }
        return HttpClient::json($resp);
    }

    private static function inboxQuery(string $mailbox): string
    {
        $mb = json_encode($mailbox);
        return 'query { inbox(mailbox: ' . $mb . ') { id headerfrom subject date } }';
    }

    private static function messageQuery(string $mailbox, string $mid): string
    {
        $mb = json_encode($mailbox);
        $midS = json_encode($mid);
        return 'query { message(mailbox: ' . $mb . ', id: ' . $midS . ') { id headerfrom subject date html } }';
    }

    public static function generate(): EmailInfo
    {
        $email = self::randomLocal(10) . '@' . self::DOMAIN;
        return new EmailInfo(self::CHANNEL, $email, $email);
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email, ?string $token): array
    {
        $mailbox = self::mailbox($email);
        if ($mailbox === '') {
            throw new \InvalidArgumentException('maildrop-cc: 邮箱地址为空');
        }

        $inboxResp = self::doGraphql(self::inboxQuery($mailbox));
        $items = $inboxResp['data']['inbox'] ?? null;
        if (!is_array($items) || $items === []) {
            return [];
        }

        $out = [];
        foreach ($items as $item) {
            if (!is_array($item)) {
                continue;
            }
            $mid = (string) ($item['id'] ?? '');
            $raw = [
                'id' => $mid,
                'from' => $item['headerfrom'] ?? '',
                'subject' => $item['subject'] ?? '',
                'date' => $item['date'] ?? '',
            ];
            if ($mid !== '') {
                try {
                    $msgResp = self::doGraphql(self::messageQuery($mailbox, $mid));
                    $msg = $msgResp['data']['message'] ?? null;
                    if (is_array($msg)) {
                        $raw = [
                            'id' => $msg['id'] ?? $mid,
                            'from' => $msg['headerfrom'] ?? '',
                            'subject' => $msg['subject'] ?? '',
                            'date' => $msg['date'] ?? '',
                            'html' => $msg['html'] ?? '',
                        ];
                    }
                } catch (\Throwable) {
                }
            }
            $out[] = Normalize::email($raw, $email);
        }
        return $out;
    }
}

File: sdk/csharp/src/Providers/MailDropCc.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// maildrop.cc 渠道 — GraphQL 公共邮箱，无认证。
/// 创建邮箱：随机用户名 + @maildrop.cc（token 仅占位）。
/// 获取邮件：inbox(mailbox) 查列表，再逐封 message(mailbox,id) 补全 html 正文。
/// </summary>
public static class MailDropCc
{
    private const string Channel = "maildrop-cc";
    private const string Domain = "maildrop.cc";
    private const string GraphqlUrl = "https://api.maildrop.cc/graphql";

    private static readonly Dictionary<string, string> Headers = new()
    {
        ["Accept"] = "application/json",
        ["Content-Type"] = "application/json",
        ["Origin"] = "https://maildrop.cc",
        ["Referer"] = "https://maildrop.cc/",
        ["User-Agent"] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
    };

    private const string LocalChars = "abcdefghijklmnopqrstuvwxyz0123456789";
    private static readonly Random Rand = new();

    private static string RandomLocal(int n = 10)
    {
        var sb = new StringBuilder();
        for (var i = 0; i < n; i++) sb.Append(LocalChars[Rand.Next(LocalChars.Length)]);
        return sb.ToString();
    }

    private static string Mailbox(string email)
    {
        email = (email ?? "").Trim();
        var idx = email.IndexOf('@');
        return idx >= 0 ? email[..idx] : email;
    }

    /// <summary>JSON 字符串字面量（含引号），用于安全嵌入 GraphQL 查询</summary>
    private static string JsonStr(string s) => Json.Serialize(s);

    private static JsonObject DoGraphql(string query)
    {
        var resp = Http.Post(GraphqlUrl, Json.Serialize(new Dictionary<string, object?> { ["query"] = query }),
            "application/json", Headers);
        resp.EnsureSuccess();
        return Json.Parse(resp.Body) as JsonObject ?? new JsonObject();
    }

    /// <summary>创建临时邮箱（无需 API 调用）</summary>
    public static EmailInfo Generate()
    {
        var email = $"{RandomLocal(10)}@{Domain}";
        return new EmailInfo(Channel, email, email);
    }

    /// <summary>获取邮件列表</summary>
    public static List<Email> GetEmails(string email, string? token)
    {
        var mailbox = Mailbox(email);
        if (mailbox.Length == 0) throw new ArgumentException("maildrop-cc: 邮箱地址为空");

        var inboxQuery = $"query {{ inbox(mailbox: {JsonStr(mailbox)}) {{ id headerfrom subject date }} }}";
        var inboxResp = DoGraphql(inboxQuery);
        var items = (inboxResp["data"] as JsonObject)?["inbox"] as JsonArray;
        var result = new List<Email>();
        if (items is null) return result;

        foreach (var it in items)
        {
            if (it is not JsonObject item) continue;
            var mid = Json.Str(item, "id");

            var raw = new Dictionary<string, object?>
            {
                ["id"] = mid,
                ["from"] = Json.Str(item, "headerfrom"),
                ["subject"] = Json.Str(item, "subject"),
                ["date"] = Json.Str(item, "date"),
            };

            if (mid.Length > 0)
            {
                try
                {
                    var msgQuery = $"query {{ message(mailbox: {JsonStr(mailbox)}, id: {JsonStr(mid)}) {{ id headerfrom subject date html }} }}";
                    var msgResp = DoGraphql(msgQuery);
                    if ((msgResp["data"] as JsonObject)?["message"] is JsonObject msg)
                        raw = new Dictionary<string, object?>
                        {
                            ["id"] = string.IsNullOrEmpty(Json.Str(msg, "id")) ? mid : Json.Str(msg, "id"),
                            ["from"] = Json.Str(msg, "headerfrom"),
                            ["subject"] = Json.Str(msg, "subject"),
                            ["date"] = Json.Str(msg, "date"),
                            ["html"] = Json.Str(msg, "html"),
                        };
                }
                catch { /* 回退列表摘要 */ }
            }

            result.Add(Normalize.NormalizeEmail(raw, email));
        }
        return result;
    }
}
